### PR TITLE
Feature/use result

### LINKED ISSUE #35: Use Result instead of failwith

### PR DESCRIPTION
**Pull request by yutopp**
Resolve #35 

This PR has some BREAKING CHANGES.

- Now `Abstract_format.of_etf` returns not `t` but `t Result.t`
- `Chunk` module is renamed to `Beam`
- `ppx_deriving` is removed. Use `sexp_of_t` instead of `show`
- Use base library instead of stdlib

### CODE REVIEW

File: example/dune [hunk 1]
```diff
@@ -1,6 +1,6 @@
 (executable
  (name read_beam)
- (libraries obeam))
+  (libraries obeam base stdio))
```
Comment (amutake) [inline example/dune]: nits: inserted an unnecessary space

File: obeam.opam [hunk 2]
```diff
@@ -11,12 +11,16 @@ build: [
...
-  "ppx_deriving"  {>= "4.2"}
   "camlzip"       {>= "1.07"}
   "zarith"        {>= "1.7"}
   "dune"          {build}
-  "ounit"         {build}
+  "ppx_here"      {build}
+  "ppx_let"       {build}
+  "ppx_sexp_conv" {build}
   "bisect_ppx"    {build}
+  "ounit"         {with-test}
```
Comment (amutake) [inline obeam.opam]: :+1:

File: lib/external_term_format.ml [hunk 3]
```diff
@@ -16,51 +42,40 @@ let uncompress_form uncompressed_size buf =
...
   (* uncompress *)
   let () = Zlib.uncompress fill export in
   (* to bitstring *)
-  Buffer.sub out_mem 0 uncompressed_size |> Bitstring.bitstring_of_string
+  Buffer.sub out_mem ~pos:0 ~len:uncompressed_size |> Bytes.to_string |> Bitstring.bitstring_of_string
 
-let bitstring_printer fmt buf =
-  Format.fprintf fmt "%s" (Bitstring.string_of_bitstring buf)
+(* http://erlang.org/doc/apps/erts/erl_ext_dist.html, 2018/10/11 *)
+let rec parse buf : (t * Bitstring.t, err_t) Result.t =
```
[the remark this answers is not in the document]
Comment (yutopp) [reply lib/external_term_format.ml]: Causes compilation errors if there is no `rec`. Perhaps isn't it idiomatic code in OCaml?

File: example/dune (hunk 1 above)
Comment (yutopp) [reply example/dune]: Thanks, fixed.

File: lib/external_term_format.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (amutake) [reply lib/external_term_format.ml]: Sorry, I mean that I think the change (`let rec parse_etf = ... let parse = ...` to `let rec parse = ... and parse_etf = ...`) is unnecessary.

The below code passed compilation.

https://github.com/amutake/obeam/commit/7f2f6d7c02a5462cb7113d781acd82f17bb955c9

[the remark this answers is not in the document]
Comment (yutopp) [reply lib/external_term_format.ml]: I got it, thanks. Fixed. I intended to bring a public function definition to top...

**Comment by yutopp** [comment]
Ready to merge!

**Comment by amutake** [comment]
Thank you for the great PR!
Can I merge?

**Comment by yutopp** [comment]
Yes, please!

**Comment by amutake** [comment]
:pray:

**Pull request merged**